File: sdks/rs/tests/incident_id_test.rs
use serde_json::Value;

use loxa::{Config, ContextCarrier, EventContext, MemorySinkStore, Params, SinkConfig};

#[test]
fn incident_id_in_params() {
    let store = MemorySinkStore::new();
    let logger = loxa::New(Config::test("test-svc").with_sink(SinkConfig::Memory(store.clone())));
    let mut ctx = logger.start_event(Params::new("test.incident").with_incident_id("inc-test-001"));
    ctx.finish("success").unwrap();
    logger.emit(&ctx).unwrap();

    let events = store.events();
    assert!(!events.is_empty(), "expected at least one emitted event");
    let parsed: Value = serde_json::from_str(&events[0]).unwrap();
    assert_eq!(
        parsed["incident_id"], "inc-test-001",
        "incident_id should be present in emitted JSON"
    );
}

#[test]
fn incident_id_from_context_carrier() {
    let carrier = ContextCarrier::new()
        .with_trace_id("trace-123")
        .with_span_id("span-456")
        .with_incident_id("inc-carrier-test");

    let params = Params::new("test.incident-carrier").inherit_from_carrier(&carrier);

    let store = MemorySinkStore::new();
    let logger = loxa::New(Config::test("test-svc").with_sink(SinkConfig::Memory(store.clone())));
    let mut ctx = logger.start_event(params);
    ctx.finish("success").unwrap();
    logger.emit(&ctx).unwrap();

    let events = store.events();
    assert!(!events.is_empty(), "expected at least one emitted event");
    let parsed: Value = serde_json::from_str(&events[0]).unwrap();
    assert_eq!(
        parsed["incident_id"], "inc-carrier-test",
        "incident_id from ContextCarrier should be present in emitted JSON"
    );
}

#[test]
fn incident_id_inherited_from_event_context() {
    let parent_ctx = EventContext::new(
        "parent-svc",
        Params::new("parent.event").with_incident_id("inc-inherit-test"),
    );

    let child_params = Params::new("child.event").inherit_from(&parent_ctx);
    let child_ctx = EventContext::new("child-svc", child_params);

    assert_eq!(
        child_ctx.incident_id,
        Some("inc-inherit-test".to_string()),
        "incident_id should be inherited from parent event context"
    );
}

#[test]
fn incident_id_omitted_when_not_set() {
    let store = MemorySinkStore::new();
    let logger = loxa::New(Config::test("test-svc").with_sink(SinkConfig::Memory(store.clone())));
    let mut ctx = logger.start_event(Params::new("test.no-incident"));
    ctx.finish("success").unwrap();
    logger.emit(&ctx).unwrap();

    let events = store.events();
    assert!(!events.is_empty(), "expected at least one emitted event");
    let parsed: Value = serde_json::from_str(&events[0]).unwrap();
    assert_eq!(
        parsed.get("incident_id"),
        None,
        "incident_id should be absent from JSON when not set"
    );
}

File: sdks/js/tests/incident-id.test.ts
import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import { Event, MemorySink, createLoxa, IncidentID } from '../src/index.ts';

describe('IncidentID', () => {
  it('sets incidentId from params', () => {
    const ev = new Event({
      event: 'test.incident',
      service: 'test-svc',
      incidentId: 'inc-test-001',
    }, 'test-svc', 'development');
    assert.equal(ev.incidentId, 'inc-test-001');
  });

  it('emits incident_id in JSON output', async () => {
    const sink = new MemorySink();
    const logger = createLoxa({
      service: 'test-svc',
      environment: 'test',
      sink,
    });

    const ctx = logger.startEvent({
      event: 'test.incident',
      kind: 'log',
      incidentId: 'inc-emit-test',
    });

    logger.finish(ctx, 'success');
    await logger.emit(ctx);

    const events = sink.getEvents();
    assert.ok(events.length > 0);
    const raw = events[0];
    const payload = typeof raw === 'string' ? JSON.parse(raw) : raw;
    assert.equal(payload.incident_id, 'inc-emit-test');
  });

  it('omits incident_id when not set', async () => {
    const sink = new MemorySink();
    const logger = createLoxa({
      service: 'test-svc',
      environment: 'test',
      sink,
    });

    const ctx = logger.startEvent({ event: 'test.no-incident', kind: 'log' });
    logger.finish(ctx, 'success');
    await logger.emit(ctx);

    const events = sink.getEvents();
    assert.ok(events.length > 0);
    const raw = events[0];
    const payload = typeof raw === 'string' ? JSON.parse(raw) : raw;
    assert.equal(payload.incident_id, undefined);
  });

  it('IncidentID attr constructor uses incident_id key', () => {
    const attr = IncidentID('inc-attr-test');
    assert.equal(attr.key, 'incident_id');
    assert.equal(attr.value, 'inc-attr-test');
  });
});
